apps/browser/bookmarks/networkfetcher.cpp

```diff
@@ -9,70 +9,77 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this file,
  * You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-#include "iconfetcher.h"
+#include "networkfetcher.h"
+#include "opensearchconfigs.h"
+
+#include <webengine.h>
 
 #include <QImage>
+#include <QUrl>
+#include <QDir>
+#include <QFile>
 
-IconFetcher::IconFetcher(QObject *parent)
+NetworkFetcher::NetworkFetcher(QObject *parent)
     : QObject(parent)
     , m_status(Null)
     , m_minimumIconSize(64) // Initial value that matches theme iconSizeMedium.
     , m_hasAcceptedTouchIcon(false)
+    , m_url(NULL)
 {
 }
 
-void IconFetcher::fetch(const QString &iconUrl)
+void NetworkFetcher::fetch(const QString &iconUrl)
 {
     updateAcceptedTouchIcon(false);
-    QUrl url(iconUrl);
-    QString path = url.path();
+    m_url = new QUrl(iconUrl);
+    QString path = m_url->path();
     updateStatus(Fetching);
     if (path.endsWith(".ico") || iconUrl.isEmpty()) {
         m_data = defaultIcon();
         updateStatus(Ready);
         emit dataChanged();
     } else {
-        QNetworkRequest request(url);
+        QNetworkRequest request(*m_url);
         QNetworkReply *reply = m_networkAccessManager.get(request);
-        connect(reply, &QNetworkReply::finished, this, &IconFetcher::dataReady);
+        connect(reply, &QNetworkReply::finished, this, &NetworkFetcher::dataReady);
         // qOverload(T functionPointer) would be handy to resolve right error method but it is introduced only
         // in Qt5.7. QNetWorkReply has signal error(QNetworkReply::NetworkError) and method error().
-        // connect(reply, qOverload<QNetworkReply::NetworkError>(&QNetworkReply::error), this, IconFetcher::error);
+        // connect(reply, qOverload<QNetworkReply::NetworkError>(&QNetworkReply::error), this, NetworkFetcher::error);
         connect(reply, SIGNAL(error(QNetworkReply::NetworkError)), this, SLOT(error(QNetworkReply::NetworkError)));
     }
 }
 
-IconFetcher::Status IconFetcher::status() const
+NetworkFetcher::Status NetworkFetcher::status() const
 {
     return m_status;
 }
 
-QString IconFetcher::data() const
+QString NetworkFetcher::data() const
 {
     return m_data;
 }
 
-QString IconFetcher::defaultIcon() const
+QString NetworkFetcher::defaultIcon() const
 {
     return DEFAULT_DESKTOP_BOOKMARK_ICON;
 }
 
-bool IconFetcher::hasAcceptedTouchIcon()
+bool NetworkFetcher::hasAcceptedTouchIcon()
 {
     return m_hasAcceptedTouchIcon;
 }
 
-void IconFetcher::dataReady()
+void NetworkFetcher::dataReady()
 {
     QNetworkReply *reply = qobject_cast<QNetworkReply*>(sender());
     if (reply) {
-        QByteArray iconData = reply->readAll();
+        m_networkData = reply->readAll();
```

Comment:
Commented on other block already but writing down the idea here:

```
    QNetworkReply *reply = qobject_cast<QNetworkReply*>(sender());
    if (reply) {
        m_networkData = reply->readAll();
        bool ready = true;
        if (m_type == OpenSearch) {
            ready = saveAsSearchEngine();
        } else {
            image.loadFromData(m_networkData);
            if (image.width() < m_minimumIconSize || image.height() < m_minimumIconSize) {
                m_data = defaultIcon();
            } else {
                m_data = QString(BASE64_IMAGE).arg(QString(m_networkData.toBase64()));
            }
            updateAcceptedTouchIcon(true);
            emit dataChanged();
        }
        reply->deleteLater();
        if (ready) {
            updateStatus(Ready);
        }
    } else if (m_type == Icon) {
        m_data = defaultIcon();
        updateStatus(Ready);
        emit dataChanged();
    }
```
Currently the code tries to create QImage out of the open search xml as well.